### ISSUE TITLE
[Nitro/Unenv] cloudflare_pages build error: Could not resolve import string_decoder

### ISSUE THREAD

**Issue by andrevandal** [nitrojs/nitro#1821]
### Environment

- `node`: 18.16.1
- `nuxt`: 3.7.4
- `drizzle-kit`: 0.19.13
- `@nuxtjs/tailwindcss`: 6.8.0
- `@planetscale/database`: 1.11.0
- `drizzle-orm`: 0.28.6
- `pnpm`: 8.9.0

### Reproduction

Clone the repo: https://github.com/andrevandal/ondevamos-web/tree/start-backend
Build.

I tried to build it through pnpm and npm, but both got errors.

### Describe the bug

When building the nuxt project, we got the error:

```
ℹ Building Nitro Server (preset: cloudflare_pages)                           nitro 2:18:00 PM
(node-resolve plugin) Could not resolve import "unenv/runtime/node/string_decoder/index/" in /Users/andre.vandal/projects/github.com/andrevandal/ondevamos-web/node_modules/readable-stream/lib/_stream_readable.js using exports defined in /Users/andre.vandal/projects/github.com/andrevandal/ondevamos-web/node_modules/unenv/package.json.

[nitro 2:18:03 PM]  ERROR  Error: Cannot resolve "unenv/runtime/node/string_decoder/index/" from "/Users/andre.vandal/projects/github.com/andrevandal/ondevamos-web/node_modules/readable-stream/lib/_stream_readable.js" and externals are not allowed!


undefined


[2:18:03 PM]  ERROR  Cannot resolve "unenv/runtime/node/string_decoder/index/" from "/Users/andre.vandal/projects/github.com/andrevandal/ondevamos-web/node_modules/readable-stream/lib/_stream_readable.js" and externals are not allowed!

  at Object.resolveId (node_modules/nuxt/node_modules/nitropack/dist/shared/nitro.0093b44c.mjs:1879:17)
  at async PluginDriver.hookFirstAndGetPlugin (node_modules/rollup/dist/es/shared/node-entry.js:25444:28)
  at async resolveId (node_modules/rollup/dist/es/shared/node-entry.js:24117:26)
  at async ModuleLoader.resolveId (node_modules/rollup/dist/es/shared/node-entry.js:24531:15)
  at async PluginDriver.hookFirstAndGetPlugin (node_modules/rollup/dist/es/shared/node-entry.js:25444:28)
  at async resolveId (node_modules/rollup/dist/es/shared/node-entry.js:24117:26)
  at async ModuleLoader.resolveId (node_modules/rollup/dist/es/shared/node-entry.js:24531:15)
  at async node_modules/@rollup/plugin-commonjs/dist/es/index.js:781:16
  at async Promise.all (index 10)
  at async node_modules/@rollup/plugin-commonjs/dist/es/index.js:773:32
`````

### Additional context

Related issues: 
- https://github.com/unjs/unenv/issues/129
- https://github.com/nuxt/nuxt/issues/21619

**Comment by DHartenfels** [comment]
I've got the same error with "vercel-edge" Preset:

```shell
ℹ Building Nitro Server (preset: vercel-edge)                                    nitro 21:32:11
(node-resolve plugin) Could not resolve import "unenv/runtime/node/string_decoder/index/" in D:\BisschenProggen\shopifyInspector\node_modules\lambdafs\node_modules\tar-stream\node_modules\readable-stream\lib\_stream_readable.js using exports defined in D:\BisschenProggen\shopifyInspector\node_modules\unenv\package.json.

[nitro 21:32:13]  ERROR  Error: Cannot resolve "unenv/runtime/node/string_decoder/index/" from "D:\\BisschenProggen\\shopifyInspector\\node_modules\\lambdafs\\node_modules\\tar-stream\\node_modules\\readable-stream\\lib\\_stream_readable.js" and externals are not allowed!


undefined


[21:32:13]  ERROR  Cannot resolve "unenv/runtime/node/string_decoder/index/" from "D:\\BisschenProggen\\shopifyInspector\\node_modules\\lambdafs\\node_modules\\tar-stream\\node_modules\\readable-stream\\lib\\_stream_readable.js" and externals are not allowed!

  at Object.resolveId (/D:/BisschenProggen/shopifyInspector/node_modules/nitropack/dist/shared/nitro.0093b44c.mjs:1879:17)
  at async PluginDriver.hookFirstAndGetPlugin (/D:/BisschenProggen/shopifyInspector/node_modules/rollup/dist/es/shared/node-entry.js:25444:28)
  at async resolveId (/D:/BisschenProggen/shopifyInspector/node_modules/rollup/dist/es/shared/node-entry.js:24117:26)
  at async ModuleLoader.resolveId (/D:/BisschenProggen/shopifyInspector/node_modules/rollup/dist/es/shared/node-entry.js:24531:15)
  at async PluginDriver.hookFirstAndGetPlugin (/D:/BisschenProggen/shopifyInspector/node_modules/rollup/dist/es/shared/node-entry.js:25444:28)
  at async resolveId (/D:/BisschenProggen/shopifyInspector/node_modules/rollup/dist/es/shared/node-entry.js:24117:26)
  at async ModuleLoader.resolveId (/D:/BisschenProggen/shopifyInspector/node_modules/rollup/dist/es/shared/node-entry.js:24531:15)
  at async /D:/BisschenProggen/shopifyInspector/node_modules/@rollup/plugin-commonjs/dist/es/index.js:781:16
  at async Promise.all (index 10)
  at async /D:/BisschenProggen/shopifyInspector/node_modules/@rollup/plugin-commonjs/dist/es/index.js:773:32



[21:32:13]  ERROR  Cannot resolve "unenv/runtime/node/string_decoder/index/" from "D:\\BisschenProggen\\shopifyInspector\\node_modules\\lambdafs\\node_modules\\tar-stream\\node_modules\\readable-stream\\lib\\_stream_readable.js" and externals are not allowed!
```  

This is my package.json file:
```javascript
{
  "name": "nuxt-app",
  "private": true,
  "type": "module",
  "scripts": {
    "build": "nuxt build",
    "dev": "nuxt dev",
    "generate": "nuxt generate",
    "preview": "nuxt preview",
    "postinstall": "nuxt prepare"
  },
  "devDependencies": {
    "@nuxt/devtools": "latest",
    "@nuxtjs/tailwindcss": "^6.8.0",
    "nuxt": "^3.7.4",
    "unenv": "^1.7.4",
    "vue": "^3.3.4",
    "vue-router": "^4.2.5"
  },
  "dependencies": {
    "chrome-aws-lambda": "^10.1.0",
    "puppeteer": "^10.1.0",
    "puppeteer-core": "^10.4.0"
  }
}

``` 

Building on 
```shell
Building Nitro Server (preset: node-server)   
```  
is fine without any problems

**Comment by Gruce** [comment]
Same but a different module:


'''
	[error] [nitro] Error: Cannot resolve "bufferutil" from "/opt/buildhome/repo/node_modules/.pnpm/ws@8.11.0/node_modules/ws/lib/buffer-util.js" and externals are not allowed!
'''

**Comment by ManasMadrecha** [comment]
Any solution to this? I am currently on `firebase` preset (with no error) and want to try `cloudflare`. Changing the preset to `cloudflare-pages` gives this error.

**Comment by aw-0** [comment]
Currently trying to build via the [deployment guide](https://nitro.unjs.io/deploy/workers) and getting this exact same error. Any updates on this?

**Comment by megatheikal** [comment]
> Same but a different module:
> 
> ```
> [error] [nitro] Error: Cannot resolve "bufferutil" from "/opt/buildhome/repo/node_modules/.pnpm/ws@8.11.0/node_modules/ws/lib/buffer-util.js" and externals are not allowed!
> ```

try `yarn add bufferutil utf-8-validate`. it worked for me

**Comment by Gruce** [comment]
> > Same but a different module:
> > ```
> > [error] [nitro] Error: Cannot resolve "bufferutil" from "/opt/buildhome/repo/node_modules/.pnpm/ws@8.11.0/node_modules/ws/lib/buffer-util.js" and externals are not allowed!
> > ```
> 
> try `yarn add bufferutil utf-8-validate`. it worked for me

This workaround confirmed.

Thank you.

**Comment by keupsonite** [comment]
In my `nuxt.config.ts` file I have this, if it helps in 

```  nitro: {
    rollupConfig: {
      external: ["bufferutil", "utf-8-validate"],
      output: {
        generatedCode: {
          symbols: true,
        },
      },
    },
  },
```

**Comment by HassanKadhim** [comment]
> bufferutil

Same problem, nothing changed.

**Comment by HassanKadhim** [comment]
> In my `nuxt.config.ts` file I have this, if it helps
> 
> ```
>     rollupConfig: {
>       external: ["bufferutil", "utf-8-validate"],
>       output: {
>         generatedCode: {
>           symbols: true,
>         },
>       },
>     },
> ```

Same problem, nothing changed.

**Comment by andrevandal** [comment]
the problem is that `unenv` ins't converting JavaScript code to Cloudflare compatible code.

adding modules to `external` (https://github.com/unjs/nitro/issues/1821#issuecomment-1884060776) just says to rollup this module will be available in runtime, which isn't (https://github.com/unjs/nitro/issues/1821#issuecomment-1884683426).

this is the map table of how unenv *convert* some modules:
https://github.com/unjs/unenv?tab=readme-ov-file#built-in-nodejs-modules

**Comment by Swiftzerr** [comment]
Was a solution ever found to @andrevandal's problem? It seems like him and many others are experiencing the same issue with no luck.

**Comment by devbrains-com** [comment]
Same issue here unfortunately.

**Comment by larify** [comment]
Same issue here, if any fix this, @ me pls

**Comment by pi0** [comment]
Hi. We are aware and this issue is under the track.

The issue is that you are using a library that is designed for Node.js and when trying to deploy to Cloudflare pages, Nitro (while trying to fix this with unenv) is not covering some subpath compatibility.

I should spend more time investigating what exact library is incompatible and a minimal (nitro only with only problematic dependency) could help to speed this up.

**Comment by milindanath** [comment]
I get the same error, and figured out it is _**firebase-admin version 11.11.0**_ that caused the issue in my case. It occurs when you create a nitro plugin and import firebase-admin. 

Here, you can find a reproduction of the error
https://stackblitz.com/~/edit/github-znntty

How to reproduce
1. go to repo https://stackblitz.com/~/edit/github-znntty
2. run NITRO_PRESET=cloudflare-pages npm run build

**Comment by milindanath** [comment]
I got the exact same issue when importing firebase-admin inside a nitro plugin. But I figured out that it was the import that was incorrect. A minimal reproduction can be found at https://stackblitz.com/edit/github-znntty

The incorrect version was

```
import admin from 'firebase-admin';

export default defineNitroPlugin((nitro) => {
  const test = admin.initializeApp;
});
```

The errors disappeared after correcting the code to the following.
```
import { initializeApp } from 'firebase-admin/app';

export default defineNitroPlugin((nitro) => {
  const test = initializeApp;
});
```

**Comment by kylebuxton** [comment]
> I got the exact same issue when importing firebase-admin inside a nitro plugin. But I figured out that it was the import that was incorrect. A minimal reproduction can be found at https://stackblitz.com/edit/github-znntty
> 
> The incorrect version was
> 
> ```
> import admin from 'firebase-admin';
> 
> export default defineNitroPlugin((nitro) => {
>   const test = admin.initializeApp;
> });
> ```
> 
> The errors disappeared after correcting the code to the following.
> 
> ```
> import { initializeApp } from 'firebase-admin/app';
> 
> export default defineNitroPlugin((nitro) => {
>   const test = initializeApp;
> });
> ```

I made this change but it didn't work on my end, I'm still getting the same error "Cannot resolve "unenv/runtime/node/string_decoder/index/""

**Comment by Swiftzerr** [comment]
after some research the problem seems to be with
`import {getFirestore} from "firebase-admin/firestore";`
as when I remove this import, building works perfectly. 

I don't know why this could be but I will keep looking.

**Comment by AmirMechouk** [comment]
Can add the following library also to the issue list
`[nitro] Error: Cannot resolve "aws-sdk" from "/opt/atlassian/pipelines/agent/build/node_modules/@opensearch-project/opensearch/lib/aws/AwsSigv4Signer.js" and externals are not allowed!`

**Comment by Mgoeloe** [comment]
Same with KnexJS

`[error] [nitro] Error: Cannot resolve "pg" from "/opt/buildhome/repo/node_modules/knex/lib/dialects/redshift/index.js" and externals are not allowed!`

**Comment by mrbubbaz** [comment]
Same with String_decoder (?)

[error] Cannot resolve "unenv/runtime/node/string_decoder/index/" from "/opt/buildhome/repo/node_modules/readable-stream/lib/_stream_readable.js" and externals are not allowed!

**Comment by chuckntaylor** [comment]
Any chance you have set an env variable like `NODE_ENV=production`?

If you did, then it is possible that the development dependencies are not being installed by Cloudflare before the build phase. For my purposes, I did want to set `NODE_ENV` for my development and production builds and so I had to make sure that development dependencies were installed.
```bash
npm install --include=dev
```

If you are trying to run different commands for different branches, I would recommend the following:
1. Create separate build scripts in your `package.json`. Could be something like:
```json
"scripts": {
    "build": "vite build",
    "build:dev": "vite build --mode development",
  }
```

2. Create a new file at your project root `build.sh` - Make sure to update if `CF_PAGES_BRANCH` statements with the name of your git branches. I am using branches `prod` and `dev` in my example.
```bash
# !/bin/bash

if [ "$CF_PAGES_BRANCH" == "prod" ]; then
  # Run the "production" script in `package.json` on the "prod" branch
  # "prod" should be replaced with the name of your Production branch

  echo "In Production - make sure dev dependencies are installed"
  npm install --include=dev
  echo "Building for Production"
  npm run build

elif [ "$CF_PAGES_BRANCH" == "dev" ]; then
  # Run the "dev" script in `package.json` on the "dev" branch
  # "dev" should be replaced with the name of your specific branch

  echo "Building for Development"
  npm run build:dev

else
  # Else run the dev script
  echo "Unrecognized Branch: $CF_PAGES_BRANCH"
  npm run build:dev
fi
```

3. Lastly, in your Cloudflare dashboard for the project, make sure you specify the build command to be `bash build.sh`

This will make sure that your branch `prod` in my example is running `npm install --include=dev` before the build, and is installing all the packages.

**Comment by CodyBontecou** [comment]
> Hi. We are aware and this issue is under the track.
> 
> The issue is that you are using a library that is designed for Node.js and when trying to deploy to Cloudflare pages, Nitro (while trying to fix this with unenv) is not covering some subpath compatibility.
> 
> I should spend more time investigating what exact library is incompatible and a minimal (nitro only with only problematic dependency) could help to speed this up.

This ended up being the issue. My specific dependency was `node-telegram-bot-api`. After removing it, the build was successful.

Something worth noting - this error was happening even during my local build, not specific to cloudflare.

**Comment by provok-me** [comment]
Same with Netlify and Sharp:
```
[error] [nitro] Error: Cannot resolve "@img/sharp-wasm32/versions" from "/opt/build/repo/node_modules/.pnpm/sharp@0.33.4/node_modules/sharp/lib/utility.js" and externals are not allowed!
```

**Comment by zou-yu** [comment]
Any updates? I'am facing same problem with argon2:
```

[error] [nitro] Error: Cannot resolve "./argon2.win32-x64-msvc.node" from "/opt/buildhome/repo/node_modules/.pnpm/@node-rs+argon2@1.7.0/node_modules/@node-rs/argon2/index.js" and externals are not allowed!
--
```

**Comment by Pachara** [comment]
No body sloved this issue ?
I got exactly same error


[info] [nitro] Building Nuxt Nitro server (preset: `cloudflare-pages`)
--
19:06:13.652 | [plugin node-resolve] Could not resolve import "unenv/runtime/node/string_decoder/index/" in /opt/buildhome/repo/node_modules/readable-stream/lib/_stream_readable.js using exports defined in /opt/buildhome/repo/node_modules/unenv/package.json.

**Comment by Mgoeloe** [comment]
> No body sloved this issue ?
> I got exactly same error
> 
> 
> [info] [nitro] Building Nuxt Nitro server (preset: `cloudflare-pages`)
> --
> 19:06:13.652 | [plugin node-resolve] Could not resolve import "unenv/runtime/node/string_decoder/index/" in /opt/buildhome/repo/node_modules/readable-stream/lib/_stream_readable.js using exports defined in /opt/buildhome/repo/node_modules/unenv/package.json.
> 
> 



Using polyfills might work.

**Comment by Pachara** [comment]
> > No body sloved this issue ?
> > I got exactly same error
> > ## [info] [nitro] Building Nuxt Nitro server (preset: `cloudflare-pages`)
> > 19:06:13.652 | [plugin node-resolve] Could not resolve import "unenv/runtime/node/string_decoder/index/" in /opt/buildhome/repo/node_modules/readable-stream/lib/_stream_readable.js using exports defined in /opt/buildhome/repo/node_modules/unenv/package.json.
> 
> Using polyfills might work.

I try, It not working. 
No mater what we try to by pass, it not working

**Comment by ryoid** [comment]
> Hi. We are aware and this issue is under the track.
> 
> The issue is that you are using a library that is designed for Node.js and when trying to deploy to Cloudflare pages, Nitro (while trying to fix this with unenv) is not covering some subpath compatibility.
> 
> I should spend more time investigating what exact library is incompatible and a minimal (nitro only with only problematic dependency) could help to speed this up.

Encountered similar issue, was not sure if I should comment here so I made a new issue. https://github.com/unjs/nitro/issues/2631

**Comment by Dmytro-Tihunov** [comment]
faced with same, is there any updates?

**Comment by clopezpro** [comment]
> Same with Netlify (netlify_edge) and Sharp:
> 
> ```
> [error] [nitro] Error: Cannot resolve "@img/sharp-wasm32/versions" from "/opt/build/repo/node_modules/.pnpm/sharp@0.33.4/node_modules/sharp/lib/utility.js" and externals are not allowed!
> ```
Maybe it's because I use NuxtImg
````<html>
<body>
<!--StartFragment-->
14:10:46.355 | [info] .nuxt/dist/server/server.mjs                                          395.63 kB │ map: 880.46 kB
-- | --
14:10:46.356 | [info] ✓ built in 13.29s
14:10:46.358 | [success] Server built in 13306ms
14:10:46.472 | [success] [nitro] Generated public dist
14:10:46.512 | [info] [nitro] Building Nuxt Nitro server (preset: `cloudflare-pages`)
14:11:03.363 | [error] [nitro] Error: Cannot resolve "@img/sharp-wasm32/versions" from "/opt/buildhome/repo/node_modules/sharp/lib/utility.js" and externals are not allowed!
14:11:03.363 |  
14:11:03.363 |  
14:11:03.363 | undefined
14:11:03.367 | [error] Cannot resolve "@img/sharp-wasm32/versions" from "/opt/buildhome/repo/node_modules/sharp/lib/utility.js" and externals are not allowed!
14:11:03.367 | at Object.resolveId (node_modules/nitropack/dist/nitro.mjs:1970:17)
14:11:03.368 | at async PluginDriver.hookFirstAndGetPlugin (node_modules/rollup/dist/es/shared/node-entry.js:19674:28)
14:11:03.368 | at async resolveId (node_modules/rollup/dist/es/shared/node-entry.js:18355:26)
14:11:03.368 | at async ModuleLoader.resolveId (node_modules/rollup/dist/es/shared/node-entry.js:18758:15)
14:11:03.368 | at async node_modules/@rollup/plugin-commonjs/dist/es/index.js:785:16
14:11:03.368 | at async Promise.all (index 5)
14:11:03.368 | at async node_modules/@rollup/plugin-commonjs/dist/es/index.js:777:32
14:11:03.369 | at async rewriteRequireExpressionsAndGetImportBlock (node_modules/@rollup/plugin-commonjs/dist/es/index.js:1354:28)
14:11:03.369 | at async transformCommonjs (node_modules/@rollup/plugin-commonjs/dist/es/index.js:1933:23)
14:11:03.370 | at async transform (node_modules/rollup/dist/es/shared/node-entry.js:18676:16)
14:11:03.371 |  
14:11:03.371 | [error] Cannot resolve "@img/sharp-wasm32/versions" from "/opt/buildhome/repo/node_modules/sharp/lib/utility.js" and externals are not allowed!
14:11:03.599 | Failed: Error while executing user command. Exited with error code: 1
14:11:03.610 | Failed: build command exited with code: 1
14:11:05.352 | Failed: error occurred while running build command

<!--EndFragment-->
</body>
</html>
```

**Comment by alexis-fresard** [comment]
### Solution found !

Hello everyone, I found a solution to this problem that worked for me.

In my project folder. I went to the file `nuxt.config.ts` and I added the following lines
```
export default {
  alias: {
    'string_decoder': 'string_decoder/',
  },
}
```

Then I built the project and now it works perfectly for me.

**Comment by timpulver** [comment]
Same issue with [vue3-moveable](https://www.npmjs.com/package/vue3-moveable) on NuxtHub (cloudflare-pages preset):

> [info] [nitro] Building Nuxt Nitro server (preset: `cloudflare-pages`)
> [error] [nitro] Error: Cannot resolve "vue3-moveable" from "/opt/buildhome/repo/.nuxt/dist/server/_nuxt/_id_.client-BV76-_Rs.js" and externals are not allowed!

This happens with default exports, as well as named exports.

Are there any workarounds or ways how to debug what sub-dependencies are causing the problem?

**Comment by brendonmatos** [comment]
> alias: {
>     'string_decoder': 'string_decoder/',
>   },

Solved my issue! Thanks

**Comment by paulvonber** [comment]
Same problem with @google-cloud/vision

**Comment by paulvonber** [comment]
all the solutions provided didn't help. Is there anything else not posted here yet that could help?

**Comment by pi0** [comment]
@paulvonber if the solution in https://github.com/unjs/nitro/issues/1821#issuecomment-2293345856 didn't work, can you please share a reproduction? 🙏🏼

**Comment by sugoidesune** [comment]
I can deploy nuxt with 
```
ssr: false
```
but with `true` i get that error.

```
export default {
  alias: {
    'string_decoder': 'string_decoder/',
  },
}
```
This fixed it. Throws an error during render but it works.
```
▲ [WARNING] Non-relative path "string_decoder/" is not allowed when "baseUrl" is not set (did you forget a leading "./"?) [tsconfig.json]
```

In the end the worker exceeded 1MiB so i couldn't test yet.

**Comment by DavidFrings** [comment]
I have the same problem...

Here is a shortet version of my logs:
```rust
Detected the following tools from environment: nodejs@20.12.2, pnpm@9.12.2
dependencies:
	+ hammerjs 2.0.8
	+ nuxt 3.13.2
	+ prettier 3.3.3
	+ vue 3.5.11
	+ vue-router 4.4.5
	
devDependencies:
	+ @emotion/css 11.13.4
	+ @nuxtjs/i18n 9.0.0-rc.2
	+ @nuxtjs/sitemap 6.1.2
	+ @vite-pwa/nuxt 0.10.5
	+ nuxt-security 2.0.0

Executing user command: pnpm install && pnpm run format && pnpm run build --if-present

> nuxt build "--if-present"
[log] Nuxt 3.13.2 with Nitro 2.9.7
[success] Client built in 2496ms
[info] Building server...
[info] [vite v5.4.8] building SSR bundle for production...
[plugin] [unplugin-vue-i18n:directive] Sourcemap is likely to be incorrect: a plugin (unplugin-vue-i18n:directive) was used to transform files, but didn't generate a sourcemap for the transformation. Consult the plugin documentation for help
[success] Server built in 1705ms
[success] [nitro] Generated public dist

PWA v0.20.5
mode      generateSW
precache  3 entries (0.41 KiB)
files generated
 dist/sw.js
 dist/workbox-cbf83eee.js

[info] [nitro] Building Nuxt Nitro server (preset: `cloudflare_pages`)
[error] [nitro] RollupError: .nuxt/dist/server/_nuxt/prompt-BGYTzdFy.js (1:9): "stdin" is not exported by "node_modules/.pnpm/unenv@1.10.0/node_modules/unenv/runtime/node/process/index.mjs", imported by ".nuxt/dist/server/_nuxt/prompt-BGYTzdFy.js".


1: import { stdin, stdout } from "node:process";
	            ^
2: import f from "node:readline";
3: import { WriteStream } from "node:tty";

[error] .nuxt/dist/server/_nuxt/prompt-BGYTzdFy.js (1:9): "stdin" is not exported by "node_modules/.pnpm/unenv@1.10.0/node_modules/unenv/runtime/node/process/index.mjs", imported by ".nuxt/dist/server/_nuxt/prompt-BGYTzdFy.js".
	  at getRollupError (node_modules/.pnpm/rollup@4.24.0/node_modules/rollup/dist/es/shared/parseAst.js:395:41)
	  at error (node_modules/.pnpm/rollup@4.24.0/node_modules/rollup/dist/es/shared/parseAst.js:391:42)
	  at Module.error (node_modules/.pnpm/rollup@4.24.0/node_modules/rollup/dist/es/shared/node-entry.js:15535:16)
	  at Module.traceVariable (node_modules/.pnpm/rollup@4.24.0/node_modules/rollup/dist/es/shared/node-entry.js:15984:29)
	  at ModuleScope.findVariable (node_modules/.pnpm/rollup@4.24.0/node_modules/rollup/dist/es/shared/node-entry.js:13770:39)
	  at ChildScope.findVariable (node_modules/.pnpm/rollup@4.24.0/node_modules/rollup/dist/es/shared/node-entry.js:5252:38)
	  at ClassBodyScope.findVariable (node_modules/.pnpm/rollup@4.24.0/node_modules/rollup/dist/es/shared/node-entry.js:5252:38)
	  at ChildScope.findVariable (node_modules/.pnpm/rollup@4.24.0/node_modules/rollup/dist/es/shared/node-entry.js:5252:38)
	  at ChildScope.findVariable (node_modules/.pnpm/rollup@4.24.0/node_modules/rollup/dist/es/shared/node-entry.js:5252:38)
	  at FunctionScope.findVariable (node_modules/.pnpm/rollup@4.24.0/node_modules/rollup/dist/es/shared/node-entry.js:5252:38)
	  at Identifier.bind (node_modules/.pnpm/rollup@4.24.0/node_modules/rollup/dist/es/shared/node-entry.js:5035:40)
	  at AssignmentPattern.bind (node_modules/.pnpm/rollup@4.24.0/node_modules/rollup/dist/es/shared/node-entry.js:2855:23)
	  at Property.bind (node_modules/.pnpm/rollup@4.24.0/node_modules/rollup/dist/es/shared/node-entry.js:2855:23)
	  at ObjectPattern.bind (node_modules/.pnpm/rollup@4.24.0/node_modules/rollup/dist/es/shared/node-entry.js:2851:28)
	  at FunctionExpression.bind (node_modules/.pnpm/rollup@4.24.0/node_modules/rollup/dist/es/shared/node-entry.js:2851:28)
	  at MethodDefinition.bind (node_modules/.pnpm/rollup@4.24.0/node_modules/rollup/dist/es/shared/node-entry.js:2855:23)
	  at ClassBody.bind (node_modules/.pnpm/rollup@4.24.0/node_modules/rollup/dist/es/shared/node-entry.js:2851:28)
	  at ClassDeclaration.bind (node_modules/.pnpm/rollup@4.24.0/node_modules/rollup/dist/es/shared/node-entry.js:2855:23)
	  at Program.bind (node_modules/.pnpm/rollup@4.24.0/node_modules/rollup/dist/es/shared/node-entry.js:2851:28)
	  at Module.bindReferences (node_modules/.pnpm/rollup@4.24.0/node_modules/rollup/dist/es/shared/node-entry.js:15514:18)
	  at Graph.sortModules (node_modules/.pnpm/rollup@4.24.0/node_modules/rollup/dist/es/shared/node-entry.js:21019:20)
	  at Graph.build (node_modules/.pnpm/rollup@4.24.0/node_modules/rollup/dist/es/shared/node-entry.js:20922:14)
	  at async node_modules/.pnpm/rollup@4.24.0/node_modules/rollup/dist/es/shared/node-entry.js:21607:13
	  at async catchUnfinishedHookActions (node_modules/.pnpm/rollup@4.24.0/node_modules/rollup/dist/es/shared/node-entry.js:21076:16)
	  at async rollupInternal (node_modules/.pnpm/rollup@4.24.0/node_modules/rollup/dist/es/shared/node-entry.js:21602:5)
	  at async _build (node_modules/.pnpm/nitropack@2.9.7_magicast@0.3.5/node_modules/nitropack/dist/nitro.mjs:2633:20)
	  at async node_modules/.pnpm/nuxt@3.13.2_@parcel+watcher@2.4.1_@types+node@22.7.4_eslint@9.13.0_jiti@2.3.1__ioredis@5.4.1__5dzmqmjnledb6iljrfoadnkny4/node_modules/nuxt/dist/index.mjs:3630:5
	  at async build (node_modules/.pnpm/nuxt@3.13.2_@parcel+watcher@2.4.1_@types+node@22.7.4_eslint@9.13.0_jiti@2.3.1__ioredis@5.4.1__5dzmqmjnledb6iljrfoadnkny4/node_modules/nuxt/dist/index.mjs:5778:3)
	  at async Object.run (node_modules/.pnpm/nuxi@3.14.0/node_modules/nuxi/dist/chunks/build.mjs:101:5)
	  at async runCommand$1 (node_modules/.pnpm/nuxi@3.14.0/node_modules/nuxi/dist/shared/nuxi.3e201632.mjs:1648:16)
	  at async runCommand$1 (node_modules/.pnpm/nuxi@3.14.0/node_modules/nuxi/dist/shared/nuxi.3e201632.mjs:1639:11)
	  at async runMain$1 (node_modules/.pnpm/nuxi@3.14.0/node_modules/nuxi/dist/shared/nuxi.3e201632.mjs:1777:7) 

[error] .nuxt/dist/server/_nuxt/prompt-BGYTzdFy.js (1:9): "stdin" is not exported by "node_modules/.pnpm/unenv@1.10.0/node_modules/unenv/runtime/node/process/index.mjs", imported by ".nuxt/dist/server/_nuxt/prompt-BGYTzdFy.js".
Command failed with exit code 1.
```

Is it because of the ssr?

Can someone help me?

**Comment by DavidFrings** [comment]
https://github.com/nuxt-modules/sitemap/issues/380#issue-2644624822

worked for me...

**Comment by KonghaYao** [comment]
![Image](https://github.com/user-attachments/assets/b646c4d0-a557-44f5-ab90-001a6418cc2a)
It's a problem of `unenv`, we need to modify the `node_modules/unenv/packages.json` 

and add exports !!!

```json
   "./runtime/node/string_decoder/index/": {
      "types": "./runtime/node/string_decoder/index.d.ts",
      "require": "./runtime/node/string_decoder/index.cjs",
      "import": "./runtime/node/string_decoder/index.mjs"
    },
```


```json
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "require": "./dist/index.cjs",
      "import": "./dist/index.mjs"
    },
    "./runtime/mock/proxy-cjs": "./runtime/mock/proxy.cjs",
    "./runtime/mock/proxy-cjs/*": "./runtime/mock/proxy.cjs",
    "./runtime/node/string_decoder/index/": {
      "types": "./runtime/node/string_decoder/index.d.ts",
      "require": "./runtime/node/string_decoder/index.cjs",
      "import": "./runtime/node/string_decoder/index.mjs"
    },
    "./runtime/*": {
      "types": "./runtime/*.d.ts",
      "require": "./runtime/*.cjs",
      "import": "./runtime/*.mjs"
    }
  },
```
@pi0  Could `unenv` fix this and release?

**Comment by KonghaYao** [comment]
```sh
sed -i '/exports/ a\    "./runtime/node/string_decoder/index/": {\n      "types": "./runtime/node/string_decoder/index.d.ts",\n      "require": "./runtime/node/string_decoder/index.cjs",\n      "import": "./runtime/node/string_decoder/index.mjs"\n    },' node_modules/unenv/package.json
```
it's a linux scripts to auto fix this probleam

**Comment by Gonzalo-Bruna** [comment]
If I manually install unenv in my project, should that fix the issue? I am facing the same problem.

**Comment by ilya-buligin** [comment]
Any updates here? This workaround stops working, my production build fail:
```
  alias: {
    'string_decoder': 'string_decoder/',
  },
```

**Comment by fahmifitu** [comment]
![Image](https://github.com/user-attachments/assets/0cecc31f-f9e0-4722-84e9-7a3b7e1f3e44)


Same here with node-server

**Comment by plcdnl** [comment]
on vercel with vercel_edge preset is impossible to deploy. Is there any workaround?

**Comment by talentlessguy** [comment]
happens with TanStack Start as well, there's a dependency that imports `readable-stream` which imports `node:string_decoder`:

```
cannot resolve "node:string_decoder/" from "/opt/buildhome/repo/node_modules/.pnpm/readable-stream@3.6.2/node_modules/readable-stream/lib/_stream_readable.js" and externals are not allowed!
```

tried to use Vite's `resolve.alias` but to no avail

```js
resolve: {
      alias: {
        'node:string_decoder': 'string_decoder/',
        'node:string_decoder/': 'string_decoder/',
      }
    },
```

**Comment by CasperSocio** [comment]
I'm having this issue and I've rolled back changes and deployed one step at a time in order to figure out exactly when the issue appears. For me, it happens when i use `import { getFirestore } from 'firebase-admin/firestore'`.

In my Nuxt project, I added a server plugin for Firebase which works just fine:

```typescript
// <rootDir>/plugins/firebase.server.ts

import { cert, getApp, initializeApp } from 'firebase-admin/app'

/**
 * Make sure that we initialize the firebase app only once
 */
const createFirebaseApp = () => {
  const config = useRuntimeConfig()

  try {
    return getApp()
  } catch {
    const firebaseConfig = {
      credential: cert(/* ••• */),
    }
    return initializeApp(firebaseConfig)
  }
}

export default defineNuxtPlugin(() => {
  createFirebaseApp()
})
```

With this server plugin, I'm still able to deploy a preview using `pnpm dlx nuxthub deploy`.

It is only when I add an API endpoint that the build is failing.

```typescript
// <rootDir>/server/api/users/[username]/is-unique.get.ts

import { getFirestore } from 'firebase-admin/firestore'

export default defineEventHandler(async (event) => {
  const username = getRouterParam(event, 'username')

  if (!username) {
    throw createError({
      statusCode: 500,
      statusMessage: 'Unable to get username from request',
    })
  }

  try {
    const snapshot = await getFirestore()
      .collection('my-totally-made-up-collection-name')
      .where('some_property', '==', username)
      .limit(1)
      .get()
    return {
      isUnique: !!snapshot.empty
    }
  } catch (err) {
    console.error('Error fetching user:', err)
    throw createError({
      statusCode: 500,
      statusMessage: 'Error fetching user',
    })
  }
})
```

It runs locally just fine, so it's not my implementation. Someone else also mentioned issues after adding the `firebase-admin/firestore` import statement.

```
[nitro ••:••:•• AM]  ERROR  Error: Cannot resolve "node:string_decoder/" from "/Users/••••••/my-project/node_modules/.pnpm/readable-stream@3.6.2/node_modules/readable-stream/lib/_stream_readable.js" and externals are not allowed!

undefined

[••:••:•• AM]  ERROR  Cannot resolve "node:string_decoder/" from "/Users/••••••/my-project/node_modules/.pnpm/readable-stream@3.6.2/node_modules/readable-stream/lib/_stream_readable.js" and externals are not allowed!

    at Object.resolveId (node_modules/.pnpm/nitropack@2.11.8_better-sqlite3@11.9.1/node_modules/nitropack/dist/rollup/index.mjs:2009:17)
    at async PluginDriver.hookFirstAndGetPlugin (node_modules/.pnpm/rollup@4.37.0/node_modules/rollup/dist/es/shared/node-entry.js:21878:28)
    at async resolveId (node_modules/.pnpm/rollup@4.37.0/node_modules/rollup/dist/es/shared/node-entry.js:20478:26)
    at async ModuleLoader.resolveId (node_modules/.pnpm/rollup@4.37.0/node_modules/rollup/dist/es/shared/node-entry.js:20914:15)
    at async PluginDriver.hookFirstAndGetPlugin (node_modules/.pnpm/rollup@4.37.0/node_modules/rollup/dist/es/shared/node-entry.js:21878:28)
    at async resolveId (node_modules/.pnpm/rollup@4.37.0/node_modules/rollup/dist/es/shared/node-entry.js:20478:26)
    at async ModuleLoader.resolveId (node_modules/.pnpm/rollup@4.37.0/node_modules/rollup/dist/es/shared/node-entry.js:20914:15)
    at async node_modules/.pnpm/@rollup+plugin-commonjs@28.0.3_rollup@4.37.0/node_modules/@rollup/plugin-commonjs/dist/es/index.js:797:16
    at async Promise.all (index 10)
    at async node_modules/.pnpm/@rollup+plugin-commonjs@28.0.3_rollup@4.37.0/node_modules/@rollup/plugin-commonjs/dist/es/index.js:789:32 

[••:••:•• AM]  ERROR  Cannot resolve "node:string_decoder/" from "/Users/••••••/my-project/node_modules/.pnpm/readable-stream@3.6.2/node_modules/readable-stream/lib/_stream_readable.js" and externals are not allowed!

 ERROR  Command failed with exit code 1: nuxi build                          ••:••:•• AM

    at getFinalError (/Users/••••••/Library/Caches/pnpm/dlx/9207406ae17a6bb1ee91c785666c56c8d9a0bcfedfdbaa447bba84b25551b626/195f75a6b00-12d65/node_modules/.pnpm/execa@9.5.2/node_modules/execa/lib/return/final-error.js:6:9)
    at makeError (/Users/••••••/Library/Caches/pnpm/dlx/9207406ae17a6bb1ee91c785666c56c8d9a0bcfedfdbaa447bba84b25551b626/195f75a6b00-12d65/node_modules/.pnpm/execa@9.5.2/node_modules/execa/lib/return/result.js:108:16)
    at getAsyncResult (/Users/••••••/Library/Caches/pnpm/dlx/9207406ae17a6bb1ee91c785666c56c8d9a0bcfedfdbaa447bba84b25551b626/195f75a6b00-12d65/node_modules/.pnpm/execa@9.5.2/node_modules/execa/lib/methods/main-async.js:167:4)
    at handlePromise (/Users/••••••/Library/Caches/pnpm/dlx/9207406ae17a6bb1ee91c785666c56c8d9a0bcfedfdbaa447bba84b25551b626/195f75a6b00-12d65/node_modules/.pnpm/execa@9.5.2/node_modules/execa/lib/methods/main-async.js:150:17)
    at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
    at async Object.run (/Users/••••••/Library/Caches/pnpm/dlx/9207406ae17a6bb1ee91c785666c56c8d9a0bcfedfdbaa447bba84b25551b626/195f75a6b00-12d65/node_modules/.pnpm/nuxthub@0.8.11/node_modules/nuxthub/src/commands/deploy.mjs:118:7)
    at async runCommand (/Users/••••••/Library/Caches/pnpm/dlx/9207406ae17a6bb1ee91c785666c56c8d9a0bcfedfdbaa447bba84b25551b626/195f75a6b00-12d65/node_modules/.pnpm/citty@0.1.6/node_modules/citty/dist/index.mjs:316:16)
    at async runCommand (/Users/••••••/Library/Caches/pnpm/dlx/9207406ae17a6bb1ee91c785666c56c8d9a0bcfedfdbaa447bba84b25551b626/195f75a6b00-12d65/node_modules/.pnpm/citty@0.1.6/node_modules/citty/dist/index.mjs:307:11)
    at async runMain (/Users/••••••/Library/Caches/pnpm/dlx/9207406ae17a6bb1ee91c785666c56c8d9a0bcfedfdbaa447bba84b25551b626/195f75a6b00-12d65/node_modules/.pnpm/citty@0.1.6/node_modules/citty/dist/index.mjs:445:7)
```

**Comment by pi0** [comment]
#3273 (since nitro 2.11.9+) adds nested aliases for `string_decoder/`, if you can still reproduce please ping with reproduction for investigate (new issue is better to track specific cases)
